src/commands/MatrixInterfaceCommand.ts

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Mjolnir } from "../Mjolnir";
+import { ApplicationCommand, ApplicationFeature, getApplicationFeature } from "./ApplicationCommand";
+
+type CommandLookupEntry = Map<string|symbol, CommandLookupEntry|MatrixInterfaceCommand<BaseFunction>>;
+
+type BaseFunction = (...args: any) => Promise<any>;
+const FLATTENED_MATRIX_COMMANDS = new Set<MatrixInterfaceCommand<BaseFunction>>();
+const THIS_COMMAND_SYMBOL = Symbol("thisCommand");
+
+type ParserSignature<ExecutorType extends (...args: any) => Promise<any>> = (
+    this: MatrixInterfaceCommand<ExecutorType>,
+    mjolnir: Mjolnir,
+    roomId: string,
+    event: any,
+    parts: string[]) => Promise<Parameters<ExecutorType>>;
```

Comment:
I'm not entirely certain that `parts` is the right type for these. It makes it impossible to parse things that contains whitespaces. Or perhaps we'll want to add a special case for `"any series of words between quotes"` and turn them into a single part?

Reply:
Yep, this was mostly for compatibility. I'm not sure how to design the parsing properly yet but I do want this to be replaced.

What i do know though is that I am thinking of something where everything is read from a stream but we can just declare the accepting types for each argument + provide a parse function for specific arguments if they don't have common types (common meaning like a matrix room or mxid).